datashuttle/utils/folders.py

```diff
@@ -576,70 +584,156 @@ def get_local_and_central_sub_or_ses_names(
         cfg,
         sub,
         search_str,
     )
 
-    # Convert subject values to a list of increasing-by-1 integers
     all_folders = list(set(folder_names["local"] + folder_names["central"]))
 
+    (
+        max_existing_num,
+        num_value_digits,
+    ) = get_max_sub_or_ses_num_and_value_length(
+        all_folders, prefix, default_num_value_digits
+    )
+
+    # calculate next sub number
+    suggested_new_num = max_existing_num + 1
+    format_suggested_new_num = str(suggested_new_num).zfill(num_value_digits)
+
+    if return_with_prefix:
+        format_suggested_new_num = f"{prefix}-{format_suggested_new_num}"
+
+    return format_suggested_new_num
+
+
+def get_max_sub_or_ses_num_and_value_length(
+    all_folders: List[str],
+    prefix: Literal["sub", "ses"],
+    default_num_value_digits: Optional[int] = None,
+) -> Tuple[int, int]:
+    """
+    Given a list of BIDS-style folder names, find the maximum subject or
+    session value (sub or ses depending on `prefix`). Also, find the
+    number of value digits across the project, so a new suggested number
+    can be formatted consistency. If the list is empty, set the value
+    to 0 and a default number of value digits.
+
+    Parameters
+    ----------
+
+    all_folders : List[str]
+        A list of BIDS-style formatted folder names.
+
+    see `get_next_sub_or_ses_number()` for other arguments.
+
+    Returns
+    -------
+
+    max_existing_num : int
+        The largest number sub / ses value in the past list.
+
+    num_value_digits : int
+        The length of the value in all sub / ses values within the
+        passed list. If these are not consistent, an error is raised.
+
+    For example, if the project contains "sub-0001", "sub-0002" then
+    the max_existing_num will be 2 and num_value_digits 4.
+
+    """
     if len(all_folders) == 0:
-        utils.raise_error(
-            "No folders found. Cannot suggest the next number.",
-            NeuroBlueprintError,
+        max_existing_num = 0
+        assert isinstance(
+            default_num_value_digits, int
+        ), "`default_num_value_digits` must be int`"
+
+        num_value_digits = default_num_value_digits
+    else:
+        all_values_str = utils.get_values_from_bids_formatted_name(
+            all_folders,
+            prefix,
+            return_as_int=False,
         )
 
-    all_value_nums = utils.get_values_from_bids_formatted_name(
-        all_folders,
-        bids_key,
-        return_as_int=True,
-        sort=True,
-    )
+        # First get the length of bids-key value across the project
+        # (e.g. sub-003 has three values).
+        all_num_value_digits = [len(value) for value in all_values_str]
 
-    if not utils.integers_are_consecutive(all_value_nums):
-        warnings.warn(
-            f"A subject number has been skipped, "
-            f"currently used subject numbers are: {all_value_nums}",
+        if not len(set(all_num_value_digits)) == 1:
```

Comment:
sonar lint warns me to use the `!=` operator instead:
```suggestion
        if len(set(all_num_value_digits)) != 1:
```

Reply:
Thanks fixed [here](https://github.com/neuroinformatics-unit/datashuttle/pull/286/commits/abac75adfe53ccd2148f3a89a6a8635e3b47e8d6).